### PR TITLE
boot, etc.: refactor boot to have a lookup with different imps

### PR DESCRIPTION
**Pull request by chipaca**
Currently the logic of boot has a lot of per-snap-type cheks in the
exported functions. Going forwards into core20 we know these checks
are going to get more involved, so now is a good time to switch things
around so that different implementations exist for each of the
situations, and thus the checks can be done while looking up which
implementation to use, instead of in each of the functions.

### CODE REVIEW

File: boot/boot.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	// ChangeRequiresReboot returns whether a reboot is required to switch
+	// to the given OS, base or kernel snap.
+	ChangeRequiresReboot() bool
+}
+
+// Lookup figures out what the right boot set is for the given
+// arguments, and returns it.
+func Lookup(s snap.PlaceInfo, t snap.Type, onClassic bool) (DeviceBootSet, error) {
+	if t != snap.TypeOS && t != snap.TypeKernel && t != snap.TypeBase && t != snap.TypeSnapd {
+		return nil, fmt.Errorf("cannot lookup boot set with snap %q of type %q", s.SnapName(), t)
```
Comment (pedronis) [inline boot/boot.go]: I would have thought this would return nil or a detectable error, or nop implementation if the type is not meaningful

File: overlord/snapstate/backend/setup.go [hunk 2]
```diff
@@ -100,7 +106,11 @@ func (b Backend) RemoveSnapFiles(s snap.PlaceInfo, typ snap.Type, meter progress
 	if _, err := os.Lstat(snapPath); err == nil {
 		// remove the kernel assets (if any)
 		if typ == snap.TypeKernel {
```
Comment (pedronis) [inline overlord/snapstate/backend/setup.go]: I would have expected this check to go away, be folded in the lookup

File: overlord/snapstate/backend/setup.go [hunk 3]
```diff
@@ -73,13 +74,18 @@ func (b Backend) SetupSnap(snapFilePath, instanceName string, sideInfo *snap.Sid
 		return snapType, err
 	}
 
-	if s.GetType() == snap.TypeKernel {
-		if err := boot.ExtractKernelAssets(s, snapf); err != nil {
+	t := s.GetType()
+	if t == snap.TypeKernel {
```
Comment (pedronis) [inline overlord/snapstate/backend/setup.go]: same expectation

File: overlord/snapstate/backend/link.go [hunk 4]
```diff
@@ -122,7 +122,11 @@ func (b Backend) LinkSnap(info *snap.Info, model *asserts.Model, tm timings.Meas
 		switch info.InstanceName() {
 		case model.Kernel(), bootBase:
```
Comment (pedronis) [inline overlord/snapstate/backend/link.go]: same expectation

File: boot/boot.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	return fmt.Errorf("cannot set next boot on classic systems")
+}
+
+func (classicBootSet) ChangeRequiresReboot() bool { return false }
+
+var _ DeviceBootSet = classicBootSet{}
+
+// InUse checks if the given name/revision is used in the
+// boot environment
+func InUse(name string, rev snap.Revision) bool {
```
Comment (zyga) [inline boot/boot.go]: I would be more comfortable with a function that returns an error rather than logging around. It's hard to determine if this is the right layer for ignoring + ignoring errors vs bubbling them up.

File: boot/boot.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+		return nil, fmt.Errorf("cannot get boot variables: %s", err)
+	}
+
+	if m["snap_mode"] == "trying" {
+		return nil, ErrBootNameAndRevisionAgain
+	}
+
+	nameAndRevno, err := nameAndRevnoFromSnap(m[bootVar])
+	if err != nil {
+		return nil, fmt.Errorf("cannot get name and revision of boot %s: %v", errName, err)
```
Comment (zyga) [inline boot/boot.go]: This won't be i18n friendly as you cannot apply declination to the value of `errName`. I would suggest rewording this to avoid the problem.

File: boot/boot.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	return nameAndRevno, nil
+}
+
+func nameAndRevnoFromSnap(sn string) (*NameAndRevision, error) {
+	if sn == "" {
+		return nil, fmt.Errorf("unset")
```
Comment (zyga) [inline boot/boot.go]: Perhaps this is correct but the extreme brevity of this error is suspicious. Perhaps it warrants a comment?

File: boot/boot.go (hunk 5 above)
Comment (chipaca) [reply boot/boot.go]: this is not new code in this PR

File: boot/boot.go (hunk 1 above)
Comment (chipaca) [reply boot/boot.go]: hm, a nop implementation might be good. We have tests that want the error I think but I'll look into it.

File: overlord/snapstate/backend/setup.go (hunk 2 above)
Comment (chipaca) [reply overlord/snapstate/backend/setup.go]: tbh I didn't even look. I will.

File: boot/boot.go (hunk 6 above)
Comment (chipaca) [reply boot/boot.go]: there is no i18n here.
Also, not new code.

File: boot/boot.go (hunk 7 above)
Comment (chipaca) [reply boot/boot.go]: ok

**Comment by pedronis** [comment]
I looked a bit more:

* in general I think we have a check that we don't install kernels on classic. That's not something that this needs to care about.

While using a nop implementations for not relevant types is tempting I think that's too confusing for this level of operations. OTOH doing type checks both here and the caller is defeating a bit  the goal, if nothing else because grepping will still find things we want to move away from.

I will make a pass with more detailed tentative proposals. I think one issue is that the single interface conflating the kernel bits is a bit awkward.

File: boot/boot.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+)
+
+// A Device Boot Set is a collection of snaps that are involved in a
+// device's boot process.
+type DeviceBootSet interface {
```
Comment (pedronis) [inline boot/boot.go]: let's split this in two interfaces.

```
type BootParticipant interface {
   SetNextBoot() error 
   ChangeRequiresReboot() error
}

type Kernel interface {
   BootParticipant

   ExtractKernelAssets(snap.Container) error
   RemoveKernelAssets() error
}
```

File: boot/boot.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+	// the right bootable base (from the model assertion).
+	SetNextBoot() error
+	// ChangeRequiresReboot returns whether a reboot is required to switch
+	// to the given OS, base or kernel snap.
+	ChangeRequiresReboot() bool
+}
+
+// Lookup figures out what the right boot set is for the given
+// arguments, and returns it.
+func Lookup(s snap.PlaceInfo, t snap.Type, onClassic bool) (DeviceBootSet, error) {
```
Comment (pedronis) [inline boot/boot.go]: the error is only used to express that the is unexpected.  Let's make this:

`Lookup(...) (bootp BootParticipant, applicable bool)`

this will return `nil, false` for not relevant types.  And a Kernel if the snap is a kernel.

Then in the caller we can two kind of checks:

```
if applicable { ... }
```

```
if krnl, ok := bootp.(boot.Kernel); ok { ... }
```

File: overlord/snapstate/backend/link.go (hunk 4 above)
Comment (chipaca) [reply overlord/snapstate/backend/link.go]: for this one I've had to refactor things again so `Lookup` takes the `model`. No biggie.

File: boot/kernel_os.go [hunk 10]
```diff
@@ -20,80 +20,77 @@
...
-// RemoveKernelAssets removes the unpacked kernel/initrd for the given
-// kernel snap.
-func RemoveKernelAssets(s snap.PlaceInfo) error {
+type coreBootParticipant struct {
+	s snap.PlaceInfo
+	t snap.Type
+}
+
+type coreKernel struct {
+	s snap.PlaceInfo
```
Comment (pedronis) [inline boot/kernel_os.go]: naively I would have expected this to contain a coreBootParticipant ?

File: boot/kernel_os.go [hunk 11]
```diff
@@ -20,80 +20,77 @@
...
 
-// ExtractKernelAssets extracts kernel/initrd/dtb data from the given
-// kernel snap, if required, to a versioned bootloader directory so
-// that the bootloader can use it.
-func ExtractKernelAssets(s *snap.Info, snapf snap.Container) error {
-	if s.GetType() != snap.TypeKernel {
-		return fmt.Errorf("cannot extract kernel assets from snap type %q", s.GetType())
-	}
-
+func (k *coreKernel) ExtractKernelAssets(snapf snap.Container) error {
```
Comment (pedronis) [inline boot/kernel_os.go]: the order of these definitions vs the struct definitions is a bit atypical, unless I'm confused

File: boot/kernel_os.go [hunk 12]
```diff
@@ -20,80 +20,77 @@
...
-func SetNextBoot(s *snap.Info) error {
-	if release.OnClassic {
-		return fmt.Errorf("cannot set next boot on classic systems")
-	}
-
-	if s.GetType() != snap.TypeOS && s.GetType() != snap.TypeKernel && s.GetType() != snap.TypeBase {
-		return fmt.Errorf("cannot set next boot to snap %q with type %q", s.SnapName(), s.GetType())
-	}
+func (k *coreKernel) SetNextBoot() error {
+	return (&coreBootParticipant{s: k.s, t: snap.TypeKernel}).SetNextBoot()
```
Comment (pedronis) [inline boot/kernel_os.go]: this looks a bit odd, see my comment above

File: image/image.go [hunk 13]
```diff
@@ -772,16 +772,22 @@ func setBootvars(downloadedSnapsInfoForBootConfig map[string]*snap.Info, model *
...
 	snapf, err := snap.Open(snapPath)
 	if err != nil {
 		return err
 	}
 
-	if err := boot.ExtractKernelAssets(info, snapf); err != nil {
-		return err
+	// image always runs in not-on-classic mode
+	bp, applicable := boot.Lookup(info, info.GetType(), model, false)
+	if !applicable {
```
Comment (pedronis) [inline image/image.go]: you don't strictly need this check given the one after

File: overlord/snapstate/backend/setup.go [hunk 14]
```diff
@@ -73,13 +74,16 @@ func (b Backend) SetupSnap(snapFilePath, instanceName string, sideInfo *snap.Sid
 		return snapType, err
 	}
 
-	if s.GetType() == snap.TypeKernel {
-		if err := boot.ExtractKernelAssets(s, snapf); err != nil {
+	t := s.GetType()
+	// XXX: maybe look into passing the model
```
Comment (pedronis) [inline overlord/snapstate/backend/setup.go]: yes, that sounds a good idea

File: overlord/snapstate/handlers.go [hunk 15]
```diff
@@ -1159,24 +1159,32 @@ func maybeRestart(t *state.Task, info *snap.Info) {
...
 	// On core systems that use a base snap we need to restart
 	// snapd when the snapd snap changes.
 	model, err := ModelFromTask(t)
 	if err != nil {
 		logger.Noticef("cannot get model assertion: %v", model)
 		return
 	}
-	if model.Base() != "" && info.GetType() == snap.TypeSnapd {
+
+	bp, applicable := boot.Lookup(info, info.GetType(), model, release.OnClassic)
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: this will return applicable false for snapd on classic but we do want to restart there, no?, missing a tests? me confused?

Comment (chipaca) [reply overlord/snapstate/handlers.go]: if you're on classic you never make it this far down

Comment (pedronis) [reply overlord/snapstate/handlers.go]: ah,  forgot that the code is fairly repetitive here.

File: boot/boot.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+// Returns ErrBootNameAndRevisionAgain if the values are temporarily not established.
+func GetCurrentBoot(t snap.Type) (*NameAndRevision, error) {
+	var bootVar, errName string
+	switch t {
+	case snap.TypeKernel:
+		bootVar = "snap_kernel"
+		errName = "kernel"
+	case snap.TypeOS, snap.TypeBase:
+		bootVar = "snap_core"
+		errName = "snap"
```
Comment (pedronis) [inline boot/boot.go]: s/snap/base/ ?

File: boot/boot.go [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+			if base == "" || base == "core" {
+				return nil, false
+			}
+
+		}
+	}
+
+	bp = &coreBootParticipant{s: s, t: t}
+	if t == snap.TypeKernel {
+		bp = &coreKernel{bp.(*coreBootParticipant)}
```
Comment (pedronis) [inline boot/boot.go]: nitpick: I would have done

```
cbp := &coreBootParticipant{...}
bp = cbp
```

above to avoid the type assertion here.

File: boot/kernel_os.go [hunk 18]
```diff
@@ -115,11 +105,13 @@ func SetNextBoot(s *snap.Info) error {
 	})
 }
 
-// ChangeRequiresReboot returns whether a reboot is required to switch
-// to the given OS, base or kernel snap.
-func ChangeRequiresReboot(s *snap.Info) bool {
-	if s.GetType() != snap.TypeKernel && s.GetType() != snap.TypeOS && s.GetType() != snap.TypeBase {
-		return false
+func (bs *coreBootParticipant) ChangeRequiresReboot() bool {
+	// XXX: at some point ChangeRequiresReboot might be in its own
```
Comment (pedronis) [inline boot/kernel_os.go]: this a potential TODO not an XXX, s/XXX/TODO/ and put a question mark at the end?

File: overlord/snapstate/backend/setup.go (hunk 14 above)
Comment (pedronis) [reply overlord/snapstate/backend/setup.go]: s/XXX/TODO/

File: boot/boot.go [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+		if bootVar == snapFile {
+			return true
+		}
+	}
+
+	return false
+}
+
+var (
+	ErrBootNameAndRevisionAgain = errors.New("boot revision not yet established")
```
Comment (zyga) [inline boot/boot.go]: Perhaps this is existing code but I find the error text and variable name disconnected.

File: boot/boot.go (hunk 7 above)
Comment (zyga) [reply boot/boot.go]: Is there a comment?

File: boot/boot.go [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+
+	nameAndRevno, err := nameAndRevnoFromSnap(m[bootVar])
+	if err != nil {
+		return nil, fmt.Errorf("cannot get name and revision of boot %s: %v", errName, err)
+	}
+
+	return nameAndRevno, nil
+}
+
+func nameAndRevnoFromSnap(sn string) (*NameAndRevision, error) {
```
Comment (zyga) [inline boot/boot.go]: I would add a comment that documents the input format, just for simplicity of scanning that vs decoding the function body.

File: boot/kernel_os.go [hunk 21]
```diff
@@ -20,80 +20,44 @@
...
-// the right bootable base (from the model assertion).
-func SetNextBoot(s *snap.Info) error {
-	if release.OnClassic {
-		return fmt.Errorf("cannot set next boot on classic systems")
-	}
-
-	if s.GetType() != snap.TypeOS && s.GetType() != snap.TypeKernel && s.GetType() != snap.TypeBase {
-		return fmt.Errorf("cannot set next boot to snap %q with type %q", s.SnapName(), s.GetType())
-	}
+var _ BootParticipant = (*coreBootParticipant)(nil)
```
Comment (zyga) [inline boot/kernel_os.go]: I understand this is a bit of static assertion but I think it warrants a comment

File: boot/kernel_os.go [hunk 22]
```diff
@@ -20,80 +20,44 @@
...
-
-	bootloader, err := bootloader.Find()
-	if err != nil {
-		return fmt.Errorf("cannot extract kernel assets: %s", err)
-	}
-
-	// ask bootloader to extract the kernel assets if needed
-	return bootloader.ExtractKernelAssets(s, snapf)
+type coreBootParticipant struct {
+	s snap.PlaceInfo
```
Comment (zyga) [inline boot/kernel_os.go]: At some point the one letter variables become unreadable IMO.

File: boot/boot_test.go [hunk 23]
```diff
@@ -0,0 +1,326 @@
...
+package boot_test
+
+import (
+	"errors"
+	"path/filepath"
+	"testing"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/boot"
```
Comment (zyga) [inline boot/boot_test.go]: Why do we have `boot` and `bootloader` separately? I would have expected to see `boot/bootloader` or even `boot/loader`.

File: image/image.go [hunk 24]
```diff
@@ -772,16 +772,19 @@ func setBootvars(downloadedSnapsInfoForBootConfig map[string]*snap.Info, model *
 	return nil
 }
 
-func extractKernelAssets(snapPath string, info *snap.Info) error {
+func extractKernelAssets(snapPath string, info *snap.Info, model *asserts.Model) error {
```
Comment (zyga) [inline image/image.go]: Does this need to take the new model context thing or is a plain model sufficient? What if this code gets executed during remodelling?

File: overlord/snapstate/backend/link.go [hunk 25]
```diff
@@ -113,18 +113,9 @@ func (b Backend) LinkSnap(info *snap.Info, model *asserts.Model, tm timings.Meas
...
-		if model.Base() != "" {
-			bootBase = model.Base()
-		}
-		switch info.InstanceName() {
-		case model.Kernel(), bootBase:
-			// XXX: This *needs* to clean up if updateCurrentSymlinks fails
-			if err := boot.SetNextBoot(info); err != nil {
-				return err
-			}
+	if bp, applicable := boot.Lookup(info, info.GetType(), model, release.OnClassic); applicable {
```
Comment (zyga) [inline overlord/snapstate/backend/link.go]: This is nice!

File: overlord/snapstate/handlers.go [hunk 26]
```diff
@@ -1148,36 +1148,39 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
...
 
-	if release.OnClassic {
-		// ignore error here as we have no way to return to caller
-		snapdSnapInstalled, _ := isInstalled(st, "snapd")
-		if (info.GetType() == snap.TypeOS && !snapdSnapInstalled) ||
-			info.GetType() == snap.TypeSnapd {
-			t.Logf("Requested daemon restart.")
-			st.RequestRestart(state.RestartDaemon)
-		}
+	model, err := ModelFromTask(t)
```
Comment (zyga) [inline overlord/snapstate/handlers.go]: I think this answers my earlier question about models and remodelling.

File: overlord/snapstate/handlers.go [hunk 27]
```diff
@@ -1148,36 +1148,39 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
...
+		}
 		return
 	}
 
-	// On core systems that use a base snap we need to restart
-	// snapd when the snapd snap changes.
-	model, err := ModelFromTask(t)
-	if err != nil {
-		logger.Noticef("cannot get model assertion: %v", model)
+	if !((release.OnClassic && typ == snap.TypeOS) || typ == snap.TypeSnapd) {
```
Comment (zyga) [inline overlord/snapstate/handlers.go]: I think this should be a cleanly named helper:

```
func RequiresDaemonRestart(onClassic bool, snapType snap.Type) bool {
    return (onClassic && typ == snap.TypeOS) || snapType == snap.TypeSnapd
}
```

Then it is clear that it should also have some input from exec stack (we don't need to restart anything when reexec is not in use).

Then having that you can do:
```
if !RequiresDaemonRestart(release.OnClassic, typ) {
    // not interesting
   return
}
```

and this is more readable IMO as to the intent.

File: boot/boot.go [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+		return nil, false
+	}
+	if t != snap.TypeOS && t != snap.TypeKernel && t != snap.TypeBase {
+		return nil, false
+	}
+
+	if model != nil {
+		switch t {
+		case snap.TypeKernel:
+			if s.InstanceName() != model.Kernel() {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply boot/boot.go]: we need this kind of checks because while we don't let install multiple kernels, a remodeling might result in multiple kernels being present

File: boot/boot_test.go (hunk 23 above)
Comment (pedronis) [reply boot/boot_test.go]: bootloader is like systemd, is low level details of bootloaders so it merits a life of its own.
